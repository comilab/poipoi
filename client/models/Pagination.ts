export default class Pagination {
  currentPage: number
  from: number
  lastPage: number
  path: string
  perPage: number
  to: number
  total: number

  constructor (data: any) {
    this.currentPage = data.currentPage
    this.from = data.from
    this.lastPage = data.lastPage
    this.path = data.path
    this.perPage = data.perPage
    this.to = data.to
    this.total = data.total
  }
}
