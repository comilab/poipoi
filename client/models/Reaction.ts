import dayjs, { Dayjs } from 'dayjs'

export default class Reaction {
  id: number
  postId: number
  emoji: string
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  constructor (data: any) {
    this.id = data.id
    this.postId = data.postId
    this.emoji = data.emoji
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }
  }
}
