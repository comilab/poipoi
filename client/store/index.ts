import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
