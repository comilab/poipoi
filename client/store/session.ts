import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
