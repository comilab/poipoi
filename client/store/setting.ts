import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
