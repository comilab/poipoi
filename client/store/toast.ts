import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
