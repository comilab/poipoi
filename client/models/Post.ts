import dayjs, { Dayjs } from 'dayjs'

import Image from '~/models/Image'
import Reaction from '~/models/Reaction'

export type ScopeUnion = 'public'|'password'|'private'

export type RatingUnion = null|'nsfw'|'r18'

export default class Post {
  id: number
  userId: number
  title: string
  caption: string
  showImagesList: boolean
  text: string
  scope: ScopeUnion
  password: string|null
  publishStart: Dayjs|null = null
  publishEnd: Dayjs|null = null
  rating: RatingUnion
  pinned: boolean
  showThumbnail: boolean
  denyRobot: boolean|null
  enableReaction: boolean|null
  allowedEmojis: string[]|null
  deniedEmojis: string[]|null
  enableTwitterShare: boolean|null
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  url: string
  textCount: number
  actualDenyRobot: boolean
  actualEnableReaction: boolean
  actualAllowedEmojis: string[]
  actualDeniedEmojis: string[]
  actualEnableTwitterShare: boolean
  emojis: string[] = []

  images: Image[] = []
  imagesCount: number

  reactions: Reaction[] = []
  reactionsCount: number

  constructor (data: any) {
    this.id = data.id
    this.userId = data.userId
    this.title = data.title
    this.caption = data.caption
    this.showImagesList = data.showImagesList
    this.text = data.text
    this.scope = data.scope
    this.password = data.password
    if (data.publishStart) {
      this.publishStart = dayjs.utc(data.publishStart).local()
    }
    if (data.publishEnd) {
      this.publishEnd = dayjs.utc(data.publishEnd).local()
    }
    this.rating = data.rating
    this.pinned = data.pinned
    this.showThumbnail = data.showThumbnail
    this.denyRobot = data.denyRobot
    this.enableReaction = data.enableReaction
    this.allowedEmojis = data.allowedEmojis
    this.deniedEmojis = data.deniedEmojis
    this.enableTwitterShare = data.enableTwitterShare
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }

    this.url = data.url
    this.textCount = data.textCount
    this.actualDenyRobot = data.actualDenyRobot
    this.actualEnableReaction = data.actualEnableReaction
    this.actualAllowedEmojis = data.actualAllowedEmojis
    this.actualDeniedEmojis = data.actualDeniedEmojis
    this.actualEnableTwitterShare = data.actualEnableTwitterShare
    this.emojis = data.emojis

    if (data.images) {
      this.images = data.images.map((image: any) => new Image(image))
    }
    this.imagesCount = data.imagesCount

    if (data.reactions) {
      this.reactions = data.reactions.map((reaction: any) => new Reaction(reaction))
    }
    this.reactionsCount = data.reactionsCount
  }

  get path () {
    return `/posts/${this.id}`
  }

  get textPages () {
    return (this.text || '').split('[newpage]').map(page => page.trim())
  }
}
