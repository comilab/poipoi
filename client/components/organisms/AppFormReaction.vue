<template lang="pug">
.flex.flex-wrap.items-center.-mx-2
  AppButton.more-button.mx-2.mb-2(
    v-if="hasMoreReactions",
    title="古いリアクションを表示",
    color="white",
    size="sm",
    :icon="faAngleDoubleUp",
    :indeterminate="indexing",
    :disabled="indexing",
    @click="onClickMore"
  )
  AppEmoji.w-6.h-6.m-2(
    v-for="reaction in displayedReactions",
    :key="reaction.id",
    :text="reaction.emoji",
    :class="{ 'cursor-pointer': store.session.isLogin }",
    @click.native="onClickEmoji(reaction)"
  )
  AppTwemojiPicker(
    :emojis="post.emojis",
    @select="onSelect"
  )
    AppButton.ml-2(
      title="リアクションを追加",
      :icon="faPlus",
      color="gray",
      size="sm",
      :indeterminate="creating",
      :disabled="creating"
    )

  AppModal(
    v-model="showDeleteModal",
    :persistent="destroying"
  )
    template(#default="{ close }")
      .text-center.my-8 このリアクションを削除してよろしいですか？
      .flex.justify-center
        AppButton.mx-2(
          color="blue",
          :icon="faTrashAlt",
          :indeterminate="destroying",
          :disabled="destroying",
          @click="deleteReaction"
        ) 削除する
        AppButton.mx-2(
          color="gray",
          :disabled="destroying",
          @click="close"
        ) キャンセル
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, computed } from 'nuxt-composition-api'
import { faAngleDoubleUp, faPlus, faTrashAlt } from '@fortawesome/free-solid-svg-icons'

import { useStore } from '~/store'
import Post from '~/models/Post'
import Reaction from '~/models/Reaction'
import useReactionsApi from '~/composables/use-reactions-api'
import AppEmoji from '~/components/atoms/AppEmoji.vue'
import AppTwemojiPicker from '~/components/atoms/AppTwemojiPicker.vue'
import AppButton from '~/components/atoms/AppButton.vue'
import AppModal from '~/components/molecules/AppModal.vue'

export default defineComponent({
  components: {
    AppEmoji,
    AppTwemojiPicker,
    AppButton,
    AppModal
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props, { emit }) {
    const store = useStore()

    const {
      reaction,
      reactions,
      pagination,
      index,
      indexing,
      create,
      creating,
      destroy,
      destroying
    } = useReactionsApi(props.post)

    const localState = reactive({
      loadedReactions: [] as Reaction[],
      showDeleteModal: false
    })

    const hasMoreReactions = computed(() => {
      if (pagination.value) {
        return pagination.value.lastPage > 1
      }
      return props.post.reactionsCount > 100
    })

    const displayedReactions = computed(() => {
      return props.post.reactions
        .concat(localState.loadedReactions)
        .reverse()
    })

    const onClickMore = async () => {
      await index({
        lt: displayedReactions.value[0].createdAt!.toDate()
      })
      localState.loadedReactions.push(...reactions.value)
    }

    const onClickEmoji = (target: Reaction) => {
      if (!store.session.isLogin.value) {
        return
      }
      reaction.value = target
      localState.showDeleteModal = true
    }

    const deleteReaction = async () => {
      const reactionId = reaction.value!.id
      await destroy()
      const key = props.post.reactions.findIndex(r => r.id === reactionId)
      if (key >= 0) {
        props.post.reactions.splice(key, 1)
      }
      emit('deleted', props.post)
      localState.showDeleteModal = false
    }

    const onSelect = async (emoji: string) => {
      await create({
        emoji
      })
      props.post.reactions.unshift(reaction.value!)
      props.post.reactionsCount++
      emit('added', props.post)
    }

    return {
      store,
      indexing,
      creating,
      destroying,
      ...toRefs(localState),
      hasMoreReactions,
      displayedReactions,
      onClickMore,
      onClickEmoji,
      deleteReaction,
      onSelect,
      faAngleDoubleUp,
      faPlus,
      faTrashAlt
    }
  }
})
</script>

<style lang="postcss" scoped>
.more-button {
  @apply w-full;
}
</style>
