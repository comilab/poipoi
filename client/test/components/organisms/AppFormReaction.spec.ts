import { fireEvent, render, waitFor, within } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { ref } from '@vue/composition-api'
import dayjs from 'dayjs'

import AppFormReaction from '~/components/organisms/AppFormReaction.vue'
import * as store from '~/store'
import useReactionsApi from '~/composables/use-reactions-api'
import settingFactory from '~/test/factories/setting'
import postFactory from '~/test/factories/post'
import Reaction from '~/models/Reaction'
import Pagination from '~/models/Pagination'
import reactionFactory from '~/test/factories/reaction'
import paginationFactory from '~/test/factories/pagination'
import useMockedComponent from '~/test/stubs/MockedComponent'
import userFactory from '~/test/factories/user'
import Post from '~/models/Post'

jest.mock('~/composables/use-reactions-api')

describe('components/organisms/AppFormReaction', () => {
  let options: any
  let reactionsApiMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true,
        TwemojiPicker: true
      }
    }

    jest.spyOn(store, 'useStore').mockReturnValue(store.store)
    store.store.setting.setData(settingFactory.build())

    reactionsApiMock = {
      reaction: ref<Reaction|null>(null),
      reactions: ref<Reaction[]>([]),
      pagination: ref<Pagination|null>(null),
      index: jest.fn().mockResolvedValue(null),
      indexing: ref(false),
      create: jest.fn().mockResolvedValue(null),
      creating: ref(false),
      destroy: jest.fn().mockResolvedValue(null),
      destroying: ref(false)
    }
    mocked(useReactionsApi).mockReturnValue({ ...reactionsApiMock })
  })

  test('renders correctly', () => {
    store.store.setting.setData(settingFactory.build({
      postDefault: {
        denyRobotScope: [],
        enableReaction: true,
        enableTwitterShare: true,
        allowedEmojis: [],
        deniedEmojis: []
      }
    }))
    const { container } = render(AppFormReaction, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('post', () => {
      test('reactions„ÅÆÁµµÊñáÂ≠ó„ÇíÈÄÜÈ†Ü„ÅßË°®Á§∫', () => {
        options.props.post = postFactory.build({
          reactions: [
            reactionFactory.build({ emoji: 'üç£' }),
            reactionFactory.build({ emoji: 'üíØ' })
          ]
        })
        const { getAllByAltText } = render(AppFormReaction, options)

        expect(getAllByAltText(/üç£|üíØ/)).toHaveLength(2)
        expect(getAllByAltText(/üç£|üíØ/)[0]).toHaveAttribute('alt', 'üíØ')
        expect(getAllByAltText(/üç£|üíØ/)[1]).toHaveAttribute('alt', 'üç£')
      })

      test('AppTwemojiPicker„Å´emojis„ÇíÊ∏°„Åô', () => {
        const { component, propsList } = useMockedComponent({ props: ['emojis'] })
        options.stubs.AppTwemojiPicker = component
        options.props.post = postFactory.build({ emojis: ['üç£'] })
        render(AppFormReaction, options)

        expect(propsList[0].emojis).toStrictEqual(['üç£'])
      })
    })
  })

  describe('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫', () => {
    beforeEach(() => {
      options.props.post = postFactory.build({
        reactions: reactionFactory.buildList(100, { emoji: 'üíØ' }),
        reactionsCount: 101
      })
    })

    test('„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Åå100‰ª∂‰ª•‰∏ä„ÅÇ„ÇãÊôÇ„Å´Ë°®Á§∫', async () => {
      const { queryByTitle, updateProps } = render(AppFormReaction, options)

      expect(queryByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).toBeInTheDocument()

      await updateProps({
        post: postFactory.build({
          reactions: reactionFactory.buildList(100),
          reactionsCount: 100
        })
      })
      expect(queryByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).not.toBeInTheDocument()
    })

    test('pagination.lastPage„Åå1„ÅÆÂ†¥Âêà„ÅØË°®Á§∫„Åó„Å™„ÅÑ', () => {
      reactionsApiMock.pagination.value = paginationFactory.build({ lastPage: 1 })
      const { queryByTitle } = render(AppFormReaction, options)

      expect(queryByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).not.toBeInTheDocument()
    })

    test('„ÇØ„É™„ÉÉ„ÇØ„Åô„Çã„Å®index„ÇíÂÆüË°å', async () => {
      const lastReaction = options.props.post.reactions.slice(-1).pop()
      const { getByTitle } = render(AppFormReaction, options)

      await fireEvent.click(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫'))

      expect(reactionsApiMock.index).toBeCalled()
      expect(reactionsApiMock.index.mock.calls[0][0].lt).toBeTruthy()
      expect(dayjs(reactionsApiMock.index.mock.calls[0][0].lt).isSame(lastReaction.createdAt)).toBeTruthy()
    })

    test('Ë™≠Ëæº‰∏≠„ÅØdisabled, indeterminate', async () => {
      const { getByTitle } = render(AppFormReaction, options)

      expect(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).not.toHaveAttribute('disabled')
      expect(within(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).queryByRole('status')).not.toBeInTheDocument()

      reactionsApiMock.indexing.value = true

      await waitFor(() => {
        expect(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).toHaveAttribute('disabled')
        expect(within(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫')).getByRole('status')).toBeInTheDocument()
      })
    })

    test('Ë™≠„ÅøËæº„Åæ„Çå„Åü„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂÖàÈ†≠„Å´ËøΩÂä†Ë°®Á§∫', async () => {
      reactionsApiMock.reactions.value = reactionFactory.buildList(1, { emoji: 'üç£' })

      const { getByTitle, container } = render(AppFormReaction, options)
      await fireEvent.click(getByTitle('Âè§„ÅÑ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíË°®Á§∫'))

      await waitFor(() => {
        expect(container.querySelectorAll('img.emoji')).toHaveLength(101)
        expect(container.querySelectorAll('img.emoji')[0]).toHaveAttribute('alt', 'üç£')
      })
    })
  })

  describe('„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíËøΩÂä†', () => {
    let twemojiPickerEmitsList: ReturnType<typeof useMockedComponent>['emitsList']

    beforeEach(() => {
      const { component, emitsList } = useMockedComponent({ emits: ['select'] })
      options.stubs.AppTwemojiPicker = component
      twemojiPickerEmitsList = emitsList
    })

    test('ÁµµÊñáÂ≠ó„ÅåÈÅ∏Êäû„Åï„Çå„Åü„ÇâÁôªÈå≤Âá¶ÁêÜ„ÇíÂÆüË°å', () => {
      render(AppFormReaction, options)
      twemojiPickerEmitsList[0].select('üç£')

      expect(reactionsApiMock.create).toBeCalledWith({ emoji: 'üç£' })
    })

    test('ÁôªÈå≤‰∏≠„ÅØdisabled, indeterminate', async () => {
      const { getByTitle } = render(AppFormReaction, options)
      reactionsApiMock.creating.value = true

      await waitFor(() => {
        expect(getByTitle('„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíËøΩÂä†')).toHaveAttribute('disabled')
        within(getByTitle('„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíËøΩÂä†')).getByRole('status')
      })
    })

    test('Âá¶ÁêÜÁµÇ‰∫ÜÂæå„ÅØadded„Ç§„Éô„É≥„Éà„ÇíÁô∫ÁÅ´', async () => {
      const { emitted } = render(AppFormReaction, options)
      twemojiPickerEmitsList[0].select('üç£')

      await waitFor(() => {
        expect(emitted().added).toHaveLength(1)
        expect(emitted().added[0][0]).toBeInstanceOf(Post)
      })
    })

    test('Âá¶ÁêÜÁµÇ‰∫ÜÂæå„ÅØ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂÖàÈ†≠„Å´ËøΩÂä†', async () => {
      reactionsApiMock.reaction.value = reactionFactory.build({ emoji: 'üç£' })
      const { emitted } = render(AppFormReaction, options)
      twemojiPickerEmitsList[0].select('üç£')

      await waitFor(() => {
        expect(emitted().added[0][0].reactions).toHaveLength(1)
        expect(emitted().added[0][0].reactions[0].emoji).toBe('üç£')
        expect(emitted().added[0][0].reactionsCount).toBe(1)
      })
    })
  })

  describe('delete', () => {
    beforeEach(() => {
      options.props.post.reactions = reactionFactory.buildList(1, { emoji: 'üç£' })
      store.store.session.setUser(userFactory.build())
    })

    test('„É≠„Ç∞„Ç§„É≥‰∏≠„Å´„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„Çí„ÇØ„É™„ÉÉ„ÇØ„Åô„Çã„Å®Á¢∫Ë™ç„É¢„Éº„ÉÄ„É´„ÇíË°®Á§∫', async () => {
      const { getByAltText, getByText } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))

      expect(getByText('„Åì„ÅÆ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂâäÈô§„Åó„Å¶„Çà„Çç„Åó„ÅÑ„Åß„Åô„ÅãÔºü')).toBeInTheDocument()
    })

    test('„É≠„Ç∞„Ç§„É≥‰∏≠„Åß„Å™„Åë„Çå„Å∞„É¢„Éº„ÉÄ„É´„ÅØÂá∫„Å™„ÅÑ', async () => {
      store.store.session.setUser(null)
      const { getByAltText, queryByText } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))

      expect(queryByText('„Åì„ÅÆ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂâäÈô§„Åó„Å¶„Çà„Çç„Åó„ÅÑ„Åß„Åô„ÅãÔºü')).not.toBeInTheDocument()
    })

    test('„É¢„Éº„ÉÄ„É´ÂÜÖ„Åß„Ç≠„É£„É≥„Çª„É´„ÇíÊäº„Åô„Å®„É¢„Éº„ÉÄ„É´„ÇíÈñâ„Åò„Çã', async () => {
      const { getByAltText, getByRole, queryByText } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))
      await fireEvent.click(getByRole('button', { name: '„Ç≠„É£„É≥„Çª„É´' }))

      expect(queryByText('„Åì„ÅÆ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂâäÈô§„Åó„Å¶„Çà„Çç„Åó„ÅÑ„Åß„Åô„ÅãÔºü')).not.toBeInTheDocument()
    })

    test('„É¢„Éº„ÉÄ„É´ÂÜÖ„ÅßÂâäÈô§„Åô„Çã„ÇíÊäº„Åô„Å®ÂâäÈô§Âá¶ÁêÜ„ÇíÂÆüË°å', async () => {
      const { getByAltText, getByRole } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))
      await fireEvent.click(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' }))

      expect(reactionsApiMock.destroy).toBeCalled()
      expect(reactionsApiMock.reaction.value).toBe(options.props.post.reactions[0])
    })

    test('ÂâäÈô§‰∏≠„ÅØ„É¢„Éº„ÉÄ„É´ÂÜÖ„Éú„Çø„É≥„ÇíÊäº„Åõ„Å™„ÅÑ', async () => {
      const { getByAltText, getByRole } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))

      expect(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' })).not.toHaveAttribute('disabled')
      expect(getByRole('button', { name: '„Ç≠„É£„É≥„Çª„É´' })).not.toHaveAttribute('disabled')

      reactionsApiMock.destroying.value = true

      await waitFor(() => {
        expect(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' })).toHaveAttribute('disabled')
        expect(getByRole('button', { name: '„Ç≠„É£„É≥„Çª„É´' })).toHaveAttribute('disabled')
      })
    })

    test('ÂâäÈô§‰∏≠„ÅØ„É¢„Éº„ÉÄ„É´„ÇíÈñâ„Åò„Çâ„Çå„Å™„ÅÑ', async () => {
      const { getByAltText, queryByLabelText, getByRole } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))

      expect(queryByLabelText('close')).toBeInTheDocument()
      expect(getByRole('button', { name: '„Ç≠„É£„É≥„Çª„É´' })).not.toHaveAttribute('disabled')

      reactionsApiMock.destroying.value = true

      await waitFor(() => {
        expect(queryByLabelText('close')).not.toBeInTheDocument()
      })
    })

    test('Âá¶ÁêÜÁµÇ‰∫ÜÂæå„ÅØdeleted„Ç§„Éô„É≥„Éà„ÇíÁô∫ÁÅ´', async () => {
      const { getByAltText, getByRole, emitted } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))
      await fireEvent.click(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' }))

      await waitFor(() => {
        expect(emitted().deleted).toHaveLength(1)
        expect(emitted().deleted[0][0]).toBeInstanceOf(Post)
      })
    })

    test('Âá¶ÁêÜÁµÇ‰∫ÜÂæå„ÅØ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂâäÈô§', async () => {
      const { getByAltText, getByRole, emitted } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))
      await fireEvent.click(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' }))

      await waitFor(() => {
        expect(emitted().deleted[0][0].reactions).toHaveLength(0)
      })
    })

    test('Âá¶ÁêÜÁµÇ‰∫ÜÂæå„ÅØ„É¢„Éº„ÉÄ„É´„ÇíÈñâ„Åò„Çã', async () => {
      const { getByAltText, getByRole, queryByText } = render(AppFormReaction, options)

      await fireEvent.click(getByAltText('üç£'))
      await fireEvent.click(getByRole('button', { name: 'ÂâäÈô§„Åô„Çã' }))

      await waitFor(() => {
        expect(queryByText('„Åì„ÅÆ„É™„Ç¢„ÇØ„Ç∑„Éß„É≥„ÇíÂâäÈô§„Åó„Å¶„Çà„Çç„Åó„ÅÑ„Åß„Åô„ÅãÔºü')).not.toBeInTheDocument()
      })
    })
  })
})
