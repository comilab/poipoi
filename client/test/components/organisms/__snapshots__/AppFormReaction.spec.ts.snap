// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppFormReaction renders correctly 1`] = `
<div>
  <div
    class="flex flex-wrap items-center -mx-2"
  >
    <!---->
    <twemojipicker-stub
      class="twemoji-picker"
      emojidata="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      emojigroups="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      pickerwidth="500"
      searchemojisfeat="true"
    >
      <button
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none ml-2 bg-gray-300 text-gray-800 hover_bg-gray-400 text-sm px-2 py-1 square-sm rounded"
        data-testid="button"
        title="リアクションを追加"
        type="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class=""
          data-testid="text"
        />
      </button>
    </twemojipicker-stub>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;
